pslab/bus/spi.py

```diff
@@ -0,0 +1,505 @@
+
+        Raises
+        ------
+        ValueError
+            If given bits per word not supported by PSLab board.
+        """
+        return self._transfer(0, bits)
+
+    def _read_bulk(self, data_to_read: int, bits: int) -> List[int]:
+        """Read data array while transmitting zeros.
+
+        Relevent SPI channel need to be enabled first.
+
+        It is a primitive SPI method, prefered to use :meth:`SPISlave.read8_bulk`
+        and :meth:`SPISlave.read16_bulk`.
+
+        Parameters
+        ----------
+        data_to_read : int
+            Number of data to read from slave device.
+        bits : int
+            The number of bits per word.
+
+        Returns
+        -------
+        list of int
+            List of data returned by slave device.
+
+        Raises
+        ------
+        ValueError
+            If given bits per word not supported by PSLab board.
+        """
+        return self._transfer_bulk([0] * data_to_read, bits)
+
+
+class SPIMaster(SPIPrimitive):
+    """SPI bus controller.
+
+    Parameters
+    ----------
+    device : :class:`SerialHandler`, optional
+        Serial connection to PSLab device. If not provided, a new one will be
+        created.
+    """
+
+    def __init__(self, device: SerialHandler = None):
+        super().__init__(device)
+        # Reset config
+        self.set_parameters()
+
+    def set_parameters(
+        self,
+        primary_prescaler: int = PPRE,
+        secondary_prescaler: int = SPRE,
+        CKE: int = CKE,
+        CKP: int = CKP,
+        SMP: int = SMP,
+    ):
+        """Set SPI parameters.
+
+        Parameters
+        ----------
+        primary_prescaler : {0, 1, 2, 3}
+            Primary Prescaler for system clock :const:`CP.CLOCK_RATE`Hz.
+            (0,1,2,3) -> (64:1,16:1,4:1,1:1).
+        secondary_prescaler : {0, 1, 2, 3, 4, 5, 6, 7}
+            Secondary prescaler (0,1,..7) -> (8:1,7:1,..1:1).
+        CKE : {0, 1}
+            SPIx Clock Edge Select bit. Serial output data changes on transition
+            {0: from Idle clock state to active clock state,
+             1: from active clock state to Idle clock state}.
+        CKP : {0, 1}
+            Clock Polarity Select bit.
+            Idle state for clock is a {0: low, 1: high} level.
+        SMP : {0, 1}
+            Input data is sampled at the {0: end, 1: middle} of data output time.
+
+        Raises
+        ------
+        ValueError
+            If any one of arguments is not in its shown range.
+        """
+        self._set_parameters(primary_prescaler, secondary_prescaler, CKE, CKP, SMP)
+
+
+class SPISlave(SPIPrimitive):
+    """SPI slave device.
+
+    Parameters
+    ----------
+    device : :class:`SerialHandler`, optional
+        Serial connection to PSLab device. If not provided, a new one will be
+        created.
+    """
+
+    def __init__(self, device: SerialHandler = None):
+        super().__init__(device)
+
+    def transfer8(self, data: int) -> int:
```

Comment:
I think `exchange*` is a more accurate name for these methods.

Reply:
I named it `transfer*` because it is mentioned like this in documentation. Also, with reference to https://www.arduino.cc/en/Reference/SPITransfer which has same functionality as ours.

Reply:
Makes sense :+1: